src/schema/v2/imageSearch.ts

```diff
@@ -0,0 +1,71 @@
+import {
+  GraphQLFieldConfig,
+  GraphQLObjectType,
+  GraphQLNonNull,
+  GraphQLString,
+  GraphQLScalarType,
+} from "graphql"
+import { ResolverContext } from "types/graphql"
+import { GraphQLUpload } from "graphql-upload"
+
+export const ImageSearchType = new GraphQLObjectType({
+  name: "ImageSearch",
+  fields: () => ({
+    filename: {
+      description: "File name",
+      type: new GraphQLNonNull(GraphQLString),
+    },
+    mimetype: {
+      description:
+        "File MIME type. Provided by the client and can’t be trusted",
+      type: new GraphQLNonNull(GraphQLString),
+    },
+    encoding: {
+      description: "File stream transfer encoding",
+      type: new GraphQLNonNull(GraphQLString),
+    },
+  }),
+})
+
+export const ImageSearchField: GraphQLFieldConfig<void, ResolverContext> = {
+  type: ImageSearchType,
+  description: "Get an image info",
+  args: {
+    image: {
+      description: "Image file",
+      // TODO: Remove `as unknown` when "graphql" is updated to version 16.x.x
+      // Wrong TS declaration for GraphQLScalarType in node_modules/graphql/type/definition.d.ts
+      type: new GraphQLNonNull((GraphQLUpload as unknown) as GraphQLScalarType),
+    },
+  },
+  resolve: async (_root, { image }) => {
```

Comment:
**question**: Would this be a reasonable spot to check for user authentication instead of at the middleware layer?

We could either do something simple that's best off of the request header values...
```typescript
if (!meLoader) throw new Error("You need to be signed in to perform this action")
```

... or something more robust that actually ensures that the token is still valid as is done [here](https://github.com/artsy/metaphysics/blob/1d0ecaf0813f78f5bd0c602b2b3c80eca6dcf426/src/schema/v2/authenticationStatus.ts):

```typescript
if (!meLoader) throw new Error("You need to be signed in to perform this action")

try {
  await meLoader()
} catch (err) {
  throw new Error("You need to be signed in to perform this action")
}
```

Reply:
I [did this](https://github.com/artsy/metaphysics/pull/3991/commits/da8d441d7f3096799ffbf48d0093b6ed438eb72d#diff-8ca357c42a580b97146120dc9b6348eb12c0f59514e8dd4de609734e2042d247R42-R48) initially. But the disadvantage of this approach is that even an unauthorized user could upload a file (you can read a little more about it [here](https://www.notion.so/artsy/Spike-Developer-can-pass-an-image-as-an-argument-to-a-query-in-Metaphysics-29536efe82fe4743836b0bbe29ccaab4#90e03d1c3b644d5fa3de2d3421763d1c

https://www.notion.so/artsy/Spike-Developer-can-pass-an-image-as-an-argument-to-a-query-in-Metaphysics-29536efe82fe4743836b0bbe29ccaab4#a37eae0dbcf64e33ad4b677f775f9e6a

https://www.notion.so/artsy/Spike-Developer-can-pass-an-image-as-an-argument-to-a-query-in-Metaphysics-29536efe82fe4743836b0bbe29ccaab4#7324467ebc124d55abf23b257b1250a9

https://www.notion.so/artsy/Spike-Developer-can-pass-an-image-as-an-argument-to-a-query-in-Metaphysics-29536efe82fe4743836b0bbe29ccaab4#a001580801a14a6db59c40cc61ce0c7d))

Reply:
Right, that makes sense. Thanks for writing those extra details!

I think before we make a change at this middleware layer, maybe we can stress test the resolver-level auth checking and see if that might be sufficient for now? There might be other solutions like IP request throttling to explore. As is, the middleware check doesn't seem very robust. An abuser could easily satisfy the check by providing a fake value in the `x-access-token` header, right? We don't actually ensure that the token is valid at this level.

What do you think? We're still early on in this project so I think we can start with a simpler implementation, enabled in staging only, as we iterate.

Reply:
> An abuser could easily satisfy the check by providing a fake value in the x-access-token header, right?

Yes, you are absolutely right about this and for this reason I also don't quite like the approach with custom middleware

> What do you think? We're still early on in this project so I think we can start with a simpler implementation, enabled in staging only, as we iterate.

If you don't mind placing the logic in graphql resolver, then I will be only glad of it 👍

Reply:
> If you don't mind placing this logic in GraphQL resolver

Sounds good! I'd feel comfortable with that + an explicit feature flag to enable the new middleware 👍🏼